Argument validation relies on nonportable behavior in `strtoul()`

## Environment

commit: 1bea9203dae5c1dd5ebf89815d98e62f97e56be5
```
Mon Mar 17 09:51:26 PM UTC 2025
ION-OPEN-SOURCE-4.1.3s
Python 3.12.8
Linux x86_64
```

## Description

Validation tests for out of range profile ID and topic ID assert equality for locale-dependent error messages. E.g.,
https://github.com/250MHz/bpmail/blob/1bea9203dae5c1dd5ebf89815d98e62f97e56be5/test/test_bpmail.py#L85

On my Fedora machine, this assertion fails because the error message is different:
```
>       assert b'strtoul(): Result too large' in send.stderr
E       AssertionError: assert b'strtoul(): Result too large' in b'strtoul(): Numerical result out of range\n'
```

We also currently rely on `errno` being set when `strtoul()` couldn't perform a conversion which does not work on my Fedora machine. For example, when checking the profile ID: https://github.com/250MHz/bpmail/blob/1bea9203dae5c1dd5ebf89815d98e62f97e56be5/src/bpmailsend.c#L149-L154
We should pass an actual pointer for `endptr` instead of `NULL` and see if it's equal to `nptr`. If it is, then no conversion was performed.


## Test

Build the project and run the regression tests:
```
$ meson setup build
$ meson compile -C build
$ meson test -C build
```

Logs show three failures:
```
=========================== short test summary info ============================
FAILED ../test/test_bpmail.py::test_send_profile_id_validation - AssertionErr...
FAILED ../test/test_bpmail.py::test_send_topic_id_validation - AssertionError...
FAILED ../test/test_bpmail.py::test_recv_topic_id_validation - AssertionError...
========================= 3 failed, 7 passed in 24.19s =========================
```

--- a/src/bpmailrecv.c
+++ b/src/bpmailrecv.c
@@ -66,7 +66,11 @@ int main(int argc, char **argv) {
         switch (ch) {
             case 't':
                 errno = 0;
-                unsigned long tflag = strtoul(optarg, NULL, 0);
+                char *endptr;
+                unsigned long tflag = strtoul(optarg, &endptr, 0);
+                if (optarg == endptr) {
+                    errno = EINVAL;
+                }
                 if (errno != 0) {
                     perror("strtoul()");
                     exit(EXIT_FAILURE);

--- a/src/bpmailsend.c
+++ b/src/bpmailsend.c
@@ -119,12 +119,16 @@ static int bpmailsend(void) {
 
 int main(int argc, char **argv) {
     int ch;
+    char *endptr;
     unsigned int topic_id = 25;
     while ((ch = getopt(argc, argv, "t:")) != -1) {
         switch (ch) {
             case 't':
                 errno = 0;
-                unsigned long tflag = strtoul(optarg, NULL, 0);
+                unsigned long tflag = strtoul(optarg, &endptr, 0);
+                if (optarg == endptr) {
+                    errno = EINVAL;
+                }
                 if (errno != 0) {
                     perror("strtoul()");
                     exit(EXIT_FAILURE);
@@ -147,7 +151,10 @@ int main(int argc, char **argv) {
     }
 
     errno = 0;
-    unsigned long _profile_id = strtoul(argv[0], NULL, 0);
+    unsigned long _profile_id = strtoul(argv[0], &endptr, 0);
+    if (argv[0] == endptr) {
+        errno = EINVAL;
+    }
     if (errno != 0) {
         perror("strtoul()");
         exit(EXIT_FAILURE);

--- a/test/test_bpmail.py
+++ b/test/test_bpmail.py
@@ -82,7 +82,7 @@ def test_send_missing_args():
 def test_send_profile_id_validation():
     send = run_bpmailsend(str(2**65), dest_eid, check=False)
     assert send.returncode != 0
-    assert b'strtoul(): Result too large' in send.stderr
+    assert b'strtoul()' in send.stderr
 
     send = run_bpmailsend(str(2**33), dest_eid, check=False)
     assert send.returncode != 0
@@ -90,13 +90,13 @@ def test_send_profile_id_validation():
 
     send = run_bpmailsend('blah', dest_eid, check=False)
     assert send.returncode != 0
-    assert b'strtoul(): Invalid argument' in send.stderr
+    assert b'strtoul()' in send.stderr
 
 
 def test_send_topic_id_validation():
     send = run_bpmailsend(f'-t {2**65}', check=False)
     assert send.returncode != 0
-    assert b'strtoul(): Result too large' in send.stderr
+    assert b'strtoul()' in send.stderr
 
     send = run_bpmailsend(f'-t {2**33}', check=False)
     assert send.returncode != 0
@@ -104,13 +104,13 @@ def test_send_topic_id_validation():
 
     send = run_bpmailsend('-t blah', check=False)
     assert send.returncode != 0
-    assert b'strtoul(): Invalid argument' in send.stderr
+    assert b'strtoul()' in send.stderr
 
 
 def test_recv_topic_id_validation():
     recv = run_bpmailrecv(f'-t {2**65}', check=False)
     assert recv.returncode != 0
-    assert b'strtoul(): Result too large' in recv.stderr
+    assert b'strtoul()' in recv.stderr
 
     recv = run_bpmailrecv(f'-t {2**33}', check=False)
     assert recv.returncode != 0
@@ -118,7 +118,7 @@ def test_recv_topic_id_validation():
 
     recv = run_bpmailrecv('-t blah', check=False)
     assert recv.returncode != 0
-    assert b'strtoul(): Invalid argument' in recv.stderr
+    assert b'strtoul()' in recv.stderr
 
 
 def test_recv_extra_args():

